agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java

```diff
@@ -0,0 +1,502 @@
+
+    @Override
+    public String getProtocolName() {
+        return PROTOCOL_NAME;
+    }
+
+    @Override
+    public String getProtocolDisplayName() {
+        return PROTOCOL_DISPLAY_NAME;
+    }
+
+    @Override
+    public String getVersion() {
+        return PROTOCOL_VERSION;
+    }
+
+    @Override
+    public List<AbstractArtnetLight> getLightMemory() { return artnetLightMemory; }
+
+    @Override
+    protected List<MetaItemDescriptor> getProtocolConfigurationMetaItemDescriptors() {
+        return PROTOCOL_META_ITEM_DESCRIPTORS;
+    }
+
+    @Override
+    protected List<MetaItemDescriptor> getLinkedAttributeMetaItemDescriptors() {
+        return ATTRIBUTE_META_ITEM_DESCRIPTORS;
+    }
+
+    @Override
+    public AssetAttribute getProtocolConfigurationTemplate() {
+        return super.getProtocolConfigurationTemplate()
+                .addMeta(
+                        new MetaItem(META_PROTOCOL_HOST, null),
+                        new MetaItem(META_PROTOCOL_PORT, null)
+                );
+    }
+
+    @Override
+    protected void doUnlinkProtocolConfiguration(AssetAttribute protocolConfiguration) {
+        synchronized (protocolMessageConsumers) {
+            protocolMessageConsumers.remove(protocolConfiguration.getReferenceOrThrow());
+        }
+        super.doUnlinkProtocolConfiguration(protocolConfiguration);
+    }
+
+    @Override
+    protected Supplier<ChannelHandler[]> getEncoderDecoderProvider(UdpIoClient<ArtnetPacket> client, AssetAttribute protocolConfiguration) {
+        Supplier<ChannelHandler[]> encoderDecoderProvider = () -> {
+            List<ChannelHandler> encodersDecoders = new ArrayList<>();
+            encodersDecoders.add(new AbstractNettyIoClient.MessageToByteEncoder<ArtnetPacket>(ArtnetPacket.class, client, new BiConsumer<ArtnetPacket, ByteBuf>() {
+                @Override
+                public void accept(ArtnetPacket packet, ByteBuf buf) {
+                    packet.toByteBuf(buf);
+                }
+            }));
+            return encodersDecoders.toArray(new ChannelHandler[0]);
+        };
+        return encoderDecoderProvider;
+    }
+
+    @Override
+    protected void doLinkAttribute(AssetAttribute attribute, AssetAttribute protocolConfiguration) {
+        AttributeRef protocolRef = protocolConfiguration.getReferenceOrThrow();
+        Consumer<ArtnetPacket> messageConsumer = artnetPacket -> {};
+        synchronized (protocolMessageConsumers) {
+            protocolMessageConsumers.compute(protocolRef, (ref, consumers) -> {
+                if (consumers == null) {
+                    consumers = new ArrayList<>();
+                }
+                consumers.add(new Pair<AttributeRef, Consumer<ArtnetPacket>>(
+                        attribute.getReferenceOrThrow(),
+                        messageConsumer
+                ));
+                return consumers;
+            });
+        }
+        if(getLinkedAttribute(attribute.getReference().orElse(null)) != null) {
```

Comment:
Nested `if` statements are best avoided if possible, prefer `if not then return`